### PR TITLE
fix(notifications): Fix parameters and add tests for participants and subscriptions

### PR DESCRIPTION
**Pull request by snigdhas**
Fixes SENTRY-1662 
Fixes SENTRY-1663

### CODE REVIEW

File: src/sentry/notifications/notificationcontroller.py [hunk 1]
```diff
@@ -266,13 +265,18 @@ def _get_layered_setting_providers(
...
                             if recipient_is_user(recipient):
                                 user_or_team_settings = self._filter_providers(
-                                    settings=setting, user_id=recipient.id, type=type.value
+                                    settings=setting,
+                                    user_id=recipient.id,
+                                    type=type.value,
+                                    provider=provider.value,
                                 )
                             elif recipient_is_team(recipient):
                                 user_or_team_settings = self._filter_providers(
```
Comment (scefali) [inline src/sentry/notifications/notificationcontroller.py:282]: @snigdhas shouldn't the `provider` be an input for the team one as well?

File: src/sentry/models/groupsubscription.py [hunk 2]
```diff
@@ -187,17 +189,28 @@ def get_participants(self, group: Group) -> ParticipantMap:
...
-                for provider in providers_by_recipient[user.id]:
-                    reason = (
+                for provider, val in providers_by_recipient[user.id].items():
+                    value = NotificationSettingsOptionEnum(val)
+                    is_subcribed = (
                         subscription_option
-                        and subscription_option.reason
-                        or GroupSubscriptionReason.implicit
+                        and subscription_option.is_active
+                        and value != NotificationSettingsOptionEnum.NEVER
```
Comment (scefali) [inline src/sentry/models/groupsubscription.py:201]: @snigdhas might be better to explicitly check if it's ALWAYS or SUBSCRIBED

File: src/sentry/notifications/notificationcontroller.py [hunk 3]
```diff
@@ -446,10 +446,13 @@ def get_subscriptions_status_for_projects(
...
                         value == NotificationSettingsOptionEnum.ALWAYS for value in setting.values()
                     ),
+                    all(
+                        value == NotificationSettingsOptionEnum.NEVER for value in setting.values()
+                    ),
                 )
 
             if not has_setting:
-                subscription_status_for_projects[project] = (False, False)
+                subscription_status_for_projects[project] = (True, False, True)
```
Comment (scefali) [inline src/sentry/notifications/notificationcontroller.py:455]: @snigdhas this tuple is hard to understand. Can we make it some object with named parameters instead?

File: src/sentry/notifications/types.py [hunk 4]
```diff
@@ -349,3 +350,10 @@ class AssigneeTargetType(Enum):
     (AssigneeTargetType.TEAM.value, "Team"),
     (AssigneeTargetType.MEMBER.value, "Member"),
 ]
+
+
+@dataclass
+class SubscriptionStatus:
```
Comment (scefali) [inline src/sentry/notifications/types.py:356]: @snigdhas should this be called `GroupSubscriptionStatus`?

File: src/sentry/notifications/notificationcontroller.py [hunk 5]
```diff
@@ -441,15 +442,19 @@ def get_subscriptions_status_for_projects(
                     continue
 
                 has_setting = True
-                subscription_status_for_projects[project] = (
+
+                subscription_status_for_projects[project] = SubscriptionStatus(
```
Comment (scefali) [inline src/sentry/notifications/notificationcontroller.py:446]: @snigdhas can we pass args to it via kwargs to make it more obvious which each argument means?

File: src/sentry/api/serializers/models/group.py [hunk 6]
```diff
@@ -584,7 +588,9 @@ def _get_subscriptions(
                 project_ids=project_ids,
                 type=NotificationSettingEnum.WORKFLOW,
             )
-            query_groups = {group for group in groups if (enabled_settings[group.project_id][1])}
+            query_groups = {
+                group for group in groups if (not enabled_settings[group.project_id][2])
```
[the remark this answers is not in the document]
Comment (snigdhas) [reply src/sentry/api/serializers/models/group.py:592]: not really - the response from notification_service needs to be json serializable. We can re-map it to SubscriptionStatus like we do below, but it's not necessarily cleaner here.

[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/api/serializers/models/group.py:592]: @snigdhas can we do a dictionary then instead of a tuple?

**Review by scefali — APPROVED** [review]

File: tests/sentry/notifications/test_notificationcontroller.py [hunk 7]
```diff
@@ -531,19 +601,31 @@ def test_get_subscriptions_status_for_projects(self):
...
+        }
 
         assert controller.get_subscriptions_status_for_projects(
             project_ids=[self.project.id],
             user=self.user,
             type=NotificationSettingEnum.QUOTA,
-        ) == {self.project.id: (False, True)}
+        ) == {
+            self.project.id: GroupSubscriptionStatus(
+                is_disabled=False, is_active=True, has_only_inactive_subscriptions=False
```
Comment (scefali) [inline tests/sentry/notifications/test_notificationcontroller.py:626]: @snigdhas passing kwargs is so much better, thank you

**Pull request merged**